pkg/rtc/supervisor/participant_supervisor.go

```diff
@@ -0,0 +1,104 @@
+package supervisor
+
+import (
+	"sync"
+	"time"
+
+	"github.com/livekit/livekit-server/pkg/rtc/types"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+	"go.uber.org/atomic"
+)
+
+const (
+	monitorInterval = 3 * time.Second
+)
+
+type ParticipantSupervisorParams struct {
+	Logger logger.Logger
+}
+
+type ParticipantSupervisor struct {
+	params ParticipantSupervisorParams
+
+	lock          sync.RWMutex
+	subscriptions map[livekit.TrackID]types.OperationMonitor
+
+	isStopped atomic.Bool
+}
+
+func NewParticipantSupervisor(params ParticipantSupervisorParams) *ParticipantSupervisor {
+	p := &ParticipantSupervisor{
+		params:        params,
+		subscriptions: make(map[livekit.TrackID]types.OperationMonitor),
+	}
+
+	go p.checkState()
+
+	return p
+}
+
+func (p *ParticipantSupervisor) Stop() {
+	p.isStopped.Store(true)
+}
+
+func (p *ParticipantSupervisor) UpdateSubscription(trackID livekit.TrackID, isSubscribed bool) {
+	p.lock.Lock()
+	sm, ok := p.subscriptions[trackID]
+	if !ok {
+		sm = NewSubscriptionMonitor(SubscriptionMonitorParams{TrackID: trackID, Logger: p.params.Logger})
+		p.subscriptions[trackID] = sm
+	}
+	sm.(*SubscriptionMonitor).UpdateSubscription(isSubscribed)
+	p.lock.Unlock()
+}
+
+func (p *ParticipantSupervisor) SetSubscribedTrack(trackID livekit.TrackID, subTrack types.SubscribedTrack) {
+	p.lock.RLock()
+	sm, ok := p.subscriptions[trackID]
+	if ok {
+		sm.(*SubscriptionMonitor).SetSubscribedTrack(subTrack)
+	}
+	p.lock.RUnlock()
+}
+
+func (p *ParticipantSupervisor) ClearSubscribedTrack(trackID livekit.TrackID, subTrack types.SubscribedTrack) {
+	p.lock.RLock()
+	sm, ok := p.subscriptions[trackID]
+	if ok {
+		sm.(*SubscriptionMonitor).ClearSubscribedTrack(subTrack)
```

Comment:
Not too happy about this. Wanted to have an interface for the monitor so that the supervisor can just walk a list of `OperationMonitors` and operate on it. But, it means casting like this because each monitor will need special events. One option is to implement `OperationMonitor` interface in some existing modules, but wanted to have a cleaner implementation. Open to ideas.

Reply:
how about a method `PostEvents` with params EventType, EventParams, then each monitor use switch to process events

Reply:
Good idea @cnderrauber . Will look at that.